src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/TransportCreation/TransportCreationHandler.cs

```diff
@@ -0,0 +1,185 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+using Microsoft.Data.SqlClient.SNI;
+using Microsoft.Data.SqlClientX.Handlers.Connection;
+
+namespace Microsoft.Data.SqlClientX.Handlers.TransportCreation
+{
+    internal class TransportCreationHandler : IHandler<ConnectionHandlerContext>
+    {
+        #if NET8_0_OR_GREATER
+        private static readonly TimeSpan DefaultPollTimeout = TimeSpan.FromSeconds(30);
+        #else
+        private const int DefaultPollTimeout = 30 * 100000; // 30 seconds as microseconds
+        #endif
+
+        /// <inheritdoc />
+        public IHandler<ConnectionHandlerContext> NextHandler { get; set; }
+
+        /// <inheritdoc />
+        public async ValueTask Handle(ConnectionHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            if (context.DataSource is null)
+            {
+                context.Error = new ArgumentNullException(nameof(context));
+                return;
+            }
+
+            try
+            {
+                // @TODO: Build CoR for handling the different protocols in order
+                if (context.DataSource.ResolvedProtocol is DataSource.Protocol.TCP)
+                {
+                    context.ConnectionStream = isAsync
+                        ? await HandleTcpRequest(context, isAsync, ct).ConfigureAwait(false)
+                        : HandleTcpRequest(context, isAsync, ct).Result;
+                }
+            }
+            catch (Exception e)
+            {
+                context.Error = e;
+            }
+
+
+            if (NextHandler is not null)
+            {
+                if (isAsync)
+                {
+                    await NextHandler.Handle(context, isAsync, ct).ConfigureAwait(false);
+                }
+                else
+                {
+                    NextHandler.Handle(context, isAsync, ct).ConfigureAwait(false).GetAwaiter().GetResult();
+                }
+            }
+        }
+
+        private ValueTask<Stream> HandleNamedPipeRequest()
+        {
+            throw new NotImplementedException();
+        }
+
+        private ValueTask<Stream> HandleSharedMemoryRequest()
+        {
+            throw new NotImplementedException();
+        }
+
+        private async ValueTask<Stream> HandleTcpRequest(ConnectionHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            ct.ThrowIfCancellationRequested();
+
+            // DNS lookup
+            var ipAddresses = isAsync
+                ? await Dns.GetHostAddressesAsync(context.DataSource.ServerName, ct).ConfigureAwait(false)
+                : Dns.GetHostAddresses(context.DataSource.ServerName);
+            if (ipAddresses.Length == 0)
+            {
+                throw new Exception("Hostname did not resolve");
```

Comment:
You need to put this exception in the Request Objects Exception property

Reply:
The strings should be localizable, consider fetching it from the Strings.resx

Reply:
Nit: Using a more specific `Exception` derived class would be nice (although the idea having some way to distinguish a timeout from a failure to resolve the DNS name is great - the lack of this distinction has annoyed me for a while.)

Reply:
I asked ai for some help with this one and it suggested this solution that resolves both localization issue and more specific exception issue :)

@saurabh500, I have this caught now in the next higher layer, so there's only one instance where the error object needs to be set before returning. Is that suitable?